Source/DafnyCore/Generic/Node.cs

```diff
@@ -144,37 +145,44 @@ public virtual IEnumerable<Assumption> Assumptions(Declaration decl) {
   public ISet<INode> Visit(Func<INode, bool> beforeChildren = null, Action<INode> afterChildren = null, Action<Exception> reportError = null) {
     reportError ??= _ => { };
     beforeChildren ??= node => true;
-    afterChildren ??= node => { };
 
     var visited = new HashSet<INode>();
-    var toVisit = new LinkedList<INode>();
+    var toVisit = new LinkedList<object>();
     toVisit.AddFirst(this);
     while (toVisit.Any()) {
       var current = toVisit.First();
       toVisit.RemoveFirst();
-      if (!visited.Add(current)) {
-        continue;
-      }
-
-      if (!beforeChildren(current)) {
-        continue;
-      }
+      if (current is INode currentNode) {
+        if (!visited.Add(currentNode)) {
+          continue;
+        }
 
-      var nodeAfterChildren = toVisit.First;
-      foreach (var child in current.Children) {
-        if (child == null) {
-          reportError(new InvalidOperationException($"Object of type {current.GetType()} has null child"));
+        if (!beforeChildren(currentNode)) {
           continue;
         }
 
-        if (nodeAfterChildren == null) {
-          toVisit.AddLast(child);
-        } else {
-          toVisit.AddBefore(nodeAfterChildren, child);
+        if (afterChildren != null) {
+          void AfterNodeChildren() => afterChildren(currentNode);
+          toVisit.AddFirst((Action)AfterNodeChildren);
```

Comment:
Nice fix, funny how simple and clean the original looks while also being completely wrong. :)